### ISSUE TITLE
Change Request: Attempt to read parserName from parser

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#18645]
### ESLint version

9.6.0

### What problem do you want to solve?

Coming over from https://github.com/typescript-eslint/typescript-eslint/issues/9475: right now, in a standard `eslint.config.js`, `RuleContext`'s `parserPath` is always `undefined`:

https://github.com/eslint/eslint/blob/7c78ad9d9f896354d557f24e2d37710cf79a27bf/lib/linter/linter.js#L978

https://github.com/eslint/eslint/blob/7c78ad9d9f896354d557f24e2d37710cf79a27bf/lib/linter/linter.js#L1920-L1924

...which is unfortunate for plugins such as `@typescript-eslint/eslint-plugin` that try to let users know when they've mismatched their parser and plugin.

```plaintext
Error: Error while loading rule '@typescript-eslint/await-thenable': You have used a rule which requires parserServices to be generated. You must therefore provide a value for the "parserOptions.project" property for @typescript-eslint/parser.
Parser: undefined
Note: detected a parser other than @typescript-eslint/parser. Make sure the parser is configured to forward "parserOptions.project" to @typescript-eslint/parser.
Occurred while linting /Users/josh/repos/repros/index.ts
```

### What do you think is the correct solution?

ESLint could do a little more work here. It could try the equivalent of:

```js
const parserName = providedConfig.languageOptions?.parser?.name;
```

That would work for at least `espree` (https://github.com/eslint/espree/blob/ef88e3795fdf8595a71a37020956d05561b1f467/espree.js#L142):

> ```js
> export const name = "espree";
> ```

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
In eslintrc mode, `context.parserPath` is the absolute path to the parser module (or `"espree"` if the default parser is used).

In flat config mode, we can't know the absolute path to the parser module, so `context.parserPath` is intentionally `undefined` and should not be used. You can use `context.languageOptions.parser` to get the parser object.

**Comment by JoshuaKGoldberg** [comment]
Ah, so is `context.parserPath` deprecated then? Should it be removed in v10?

**Comment by mdjermanovic** [comment]
> Ah, so is `context.parserPath` deprecated then? Should it be removed in v10?

Yes. It's already marked as removed since it's always `undefined` in flat config mode.

https://github.com/eslint/eslint/blob/1381394a75b5902ce588455765a3919e2f138a7a/docs/src/extend/custom-rules.md?plain=1#L137

In v10, we'll completely remove the property.
